collect_app/src/main/java/org/odk/collect/android/http/OkHttpConnection.java

```diff
@@ -0,0 +1,357 @@
+
+                multipartBuilder.addPart(MultipartBody.Part.create(fileRequestBody));
+
+                byteCount += file.length();
+                Timber.i("added file of type '%s' %s", mime, file.getName());
+
+                // we've added at least one attachment to the request...
+                if (fileIndex + 1 < fileList.size()) {
+                    if ((fileIndex - lastFileIndex + 1 > 100) || (byteCount + fileList.get(fileIndex + 1).length()
+                            > 10000000L)) {
+                        // the next file would exceed the 10MB threshold...
+                        Timber.i("Extremely long post is being split into multiple posts");
+                        multipartBuilder.addPart(MultipartBody.Part.createFormData("*isIncomplete*", "yes"));
+                        ++fileIndex; // advance over the last attachment added...
+                        break;
+                    }
+                }
+            }
+
+            multipartBody = multipartBuilder.build();
+            postResult = executePostRequest(uri, credentials);
+            multipartBody = null;
+
+            if (postResult.getResponseCode() != HttpURLConnection.HTTP_CREATED &&
+                postResult.getResponseCode() != HttpURLConnection.HTTP_ACCEPTED) {
+                return postResult;
+            }
+
+        }
+
+        return postResult;
+    }
+
+    private OkHttpClient createOkHttpClient(@Nullable HttpCredentialsInterface credentials, URI uri) {
+        OkHttpClient.Builder builder;
+
+        if (httpClient != null) {
+            if (sameCredentials(credentials)) {
+                return httpClient;
+            }
+            builder = httpClient.newBuilder();
+        } else {
+            builder = new OkHttpClient.Builder();
+        }
+
+        addCredentials(builder, credentials, uri.getScheme().equalsIgnoreCase("https"));
+        httpCredentials = credentials;
+
+        httpClient = builder
+                    .connectTimeout(CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                    .writeTimeout(WRITE_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                    .readTimeout(READ_CONNECTION_TIMEOUT, TimeUnit.MILLISECONDS)
+                    .followRedirects(true)
+                    .build();
+
+        return httpClient;
+    }
+
+    private boolean sameCredentials(HttpCredentialsInterface credentials) {
+        if (httpCredentials == null && credentials == null) {
+            return true;
+        } else if (httpCredentials == null || credentials == null) {
+            return false;
+        } else if (httpCredentials.isEqualTo(credentials)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    private void addCredentials(OkHttpClient.Builder builder, @Nullable HttpCredentialsInterface credentials, Boolean sslEnabled) {
+        if (credentials != null) {
+            final Map<String, CachingAuthenticator> authCache = new ConcurrentHashMap<>();
```

Comment:
Could you please describe your understanding of the caching that occurs here? Since this cache is rebuilt every time `addCredentials` is run, I would expect it to only cache headers between requests with the same credentials. But I notice that in fact there's some unexpected behavior -- if I put in correct credentials in the server settings, do a request and then change the credentials to be wrong, the request still goes through. In fact, the latest and incorrect credentials should be used. Is this the level that caching occurs at? Somewhere else?

Reply:
I've investigated and it looks like the credentials will be cached unless you create a new http client. I've modified the code to recreate the http client if the credentials change or the scheme changes.